src/main/scala/vectorpipe/osm/internal/functions/package.scala

```diff
@@ -0,0 +1,636 @@
+      None
+    } else {
+
+      val geoms = wkbs.map { geom =>
+        geom match {
+          case p: Polygon => Some(p.getExteriorRing())
+          case l: LineString => Some(l)
+          case null => None
+        }
+      }
+
+      val vertexCount = geoms.filter(_.isDefined).map(_.get).map(_.getNumPoints).sum
+
+      logger.warn(s"${vertexCount.formatted("%,d")} vertices from ${types.size} members in $id @ $version ($timestamp)")
+
+      val members: Seq[(String, LineString)] = roles.zip(geoms)
+        .filter(_._2.isDefined)
+        .map(x => (x._1, x._2.get))
+
+      val (complete, partial) = members.foldLeft((Vector.empty[Polygon], Vector.empty[LineString])) {
+        case ((c, p), (role, line: LineString)) =>
+          role match {
+            case "outer" if line.isClosed && line.getNumPoints >= 4 =>
+              (c :+ GeomFactory.factory.createPolygon(line.getCoordinateSequence), p)
+            case "outer" => (c, p :+ line)
+            case "inner" if line.isClosed && line.getNumPoints >= 4 =>
+              (c :+ GeomFactory.factory.createPolygon(line.getCoordinateSequence), p)
+            case "inner" => (c, p :+ line)
+            case "" if line.isClosed && line.getNumPoints >= 4 =>
+              (c :+ GeomFactory.factory.createPolygon(line.getCoordinateSequence), p)
+            case "" => (c, p :+ line)
+            case _ => (c, p)
+          }
+      }
+
+      try {
+        val rings = complete ++ formRings(partial.sortWith(_.getNumPoints > _.getNumPoints))
+        val preparedRings = rings
+
+        // reclassify rings according to their topology (ignoring roles)
+        val (classifiedOuters, classifiedInners) = rings.sortWith(_.getArea > _.getArea) match {
+          case Seq(h, t@_ *) => t.foldLeft((Vector(h), Vector.empty[Polygon])) {
+            case ((os, is), ring) =>
+              // check the number of containing elements
+              preparedRings.count(r => r != ring && r.contains(ring)) % 2 match {
+                // if even, it's an outer ring
+                case 0 => (os :+ ring, is)
+                // if odd, it's an inner ring
+                case 1 => (os, is :+ ring)
+              }
+          }
+          case rs if rs.isEmpty => (Vector.empty[Polygon], Vector.empty[Polygon])
+        }
+
+        val (dissolvedOuters, addlInners) = dissolveRings(classifiedOuters)
+        val (dissolvedInners, addlOuters) = dissolveRings(classifiedInners.map(_.getExteriorRing.getCoordinateSequence).map(GeomFactory.factory.createPolygon) ++ addlInners)
+
+        val (polygons, _) = (dissolvedOuters ++ addlOuters)
+          // sort by size (descending) to use rings as part of the largest available polygon
+          .sortWith(_.getArea > _.getArea)
+          // only use inners once if they're contained by multiple outer rings
+          .foldLeft((Vector.empty[Polygon], dissolvedInners)) {
+          case ((ps, is), outer) =>
+            val preparedOuter = outer
```

Comment:
Same as the above comment where `preparedOut` is now just a `Geometry`. Original code: https://github.com/azavea/osmesa/blob/master/src/ingest/src/main/scala/osmesa/functions/osm/package.scala#L535